### PR TITLE
feat(logspout): support sending log via tcp

### PR DESCRIPTION
**Pull request by laurrentt**
I'm using AWS' Elastic Load Balancer in front on my logstash server, however ELBs don't support UDP. So here's a quick fix to add TCP support.

### CODE REVIEW

**Comment by mboersma** [comment]
ping @kalbasit, since you asked to be notified about Go code to review. :smile:

**Comment by laurrentt** [comment]
I'm a Go newbie so all comments and critics are welcome.

File: logspout/logspout.go [hunk 1]
```diff
@@ -161,6 +173,14 @@ func httpStreamer(w http.ResponseWriter, req *http.Request, logstream chan *Log,
 	}
 }
 
+func getEtcdValueOrDefault(c *etcd.Client, key string, defaultValue string) string {
+	resp, err := c.Get(key, false, false)
+	if err != nil {
```
Comment (bacongobbler) [inline logspout/logspout.go]: We might want to check the error here, as other errors like if the server are unavailable can be globbed here,

File: logspout/logspout.go [hunk 2]
```diff
@@ -183,9 +203,10 @@ func main() {
 		assert(err, "url")
 		portResp, err := etcd.Get("/deis/logs/port", false, false)
 		assert(err, "url")
+		protocol := getEtcdValueOrDefault(etcd, "/deis/logs/protocol", "upd")
```
Comment (bacongobbler) [inline logspout/logspout.go]: typo: udp

File: logspout/logspout.go [hunk 3]
```diff
@@ -161,6 +173,17 @@ func httpStreamer(w http.ResponseWriter, req *http.Request, logstream chan *Log,
 	}
 }
 
+func getEtcdValueOrDefault(c *etcd.Client, key string, defaultValue string) string {
+	resp, err := c.Get(key, false, false)
+	if err != nil {
+		if strings.Contains(fmt.Sprintf("%v", err), "Key not found") {
```
Comment (laurrentt) [inline logspout/logspout.go]: @bacongobbler: This works, let me know if it's good enough.

Comment (bacongobbler) [inline logspout/logspout.go]: +1, exactly what I was looking for

**Comment by krancour** [comment]
I haven't looked at this too closely yet, but am curious what the effort would be to use TCP over TLS so logs can (optionally) be sent securely.

**Comment by krancour** [comment]
^ Not to imply that we couldn't accept without that.

**Comment by laurrentt** [comment]
@krancour in my case I only need unsecure tcp since none of this traffic is going on the internet. 

I checked and the interface for using TLS is basically the same but it would imply adding certificate configuration, etc. 

I suggest a "Help Wanted" issue could be opened for this.

**Comment by krancour** [comment]
@laurrentt right... when/if this is accepted, we can open or seek out another PR to add TLS.

**Comment by krancour** [comment]
Related #4292 <-- may already support a variety of transport options

**Comment by laurrentt** [comment]
@krancour Yes that would definitely be a good idea. I could start using the third-party module logspout-redis which would buffer my logs in case my logstash is swamped or down.

**Comment by carmstrong** [comment]
@bacongobbler @krancour Where are we with this? Are we okay with this as-is?

**Comment by bacongobbler** [comment]
works for me until we migrate to a newer version of logspout. Code LGTM.

**Comment by kalbasit** [comment]
LGTM for the Go code.

**Pull request merged**